### ISSUE TITLE
OpenCV Members are not recognized

### ISSUE THREAD

**Issue by LambdaScorpii** [pylint-dev/pylint#10027]
### Bug description

when working with cv2 and saving files, PyLint does complain it cannot find the cv2 members, hence, flooding the Problems tab with messages, which are in fact no problem at all. Below is a MWE to reproduce some of the messages. 

```python
"""Minimum Working Example showing the issue that PyLint does not recognize cv2 Members"""

import cv2

cap = cv2.VideoCapture("test.mp4")
background_substractor = cv2.createBackgroundSubtractorMOG2()

ret, frame = cap.read()
foreground_mask = background_substractor.apply(frame)
contours, hierarchy = cv2.findContours(
    foreground_mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE
)

cv2.destroyAllWindows()
cap.release()
```

### Command used

VSCode PyLint Extension
```console
pylint --reports=n --output-format=json --clear-cache-post-run=y --from-stdin opencv_mwe_pylint_err.py
```
### Pylint output

[
    {
        "type": "error",
        "module": "object_tracker.opencv_mwe_pylint_err",
        "obj": "",
        "line": 5,
        "column": 6,
        "endLine": 5,
        "endColumn": 22,
        "path": "poc\\object_tracker\\opencv_mwe_pylint_err.py",
        "symbol": "no-member",
        "message": "Module 'cv2' has no 'VideoCapture' member",
        "message-id": "E1101"
    },
    {
        "type": "error",
        "module": "object_tracker.opencv_mwe_pylint_err",
        "obj": "",
        "line": 6,
        "column": 25,
        "endLine": 6,
        "endColumn": 59,
        "path": "poc\\object_tracker\\opencv_mwe_pylint_err.py",
        "symbol": "no-member",
        "message": "Module 'cv2' has no 'createBackgroundSubtractorMOG2' member",
        "message-id": "E1101"
    },
    {
        "type": "error",
        "module": "object_tracker.opencv_mwe_pylint_err",
        "obj": "",
        "line": 11,
        "column": 22,
        "endLine": 11,
        "endColumn": 38,
        "path": "poc\\object_tracker\\opencv_mwe_pylint_err.py",
        "symbol": "no-member",
        "message": "Module 'cv2' has no 'findContours' member",
        "message-id": "E1101"
    },
    {
        "type": "error",
        "module": "object_tracker.opencv_mwe_pylint_err",
        "obj": "",
        "line": 12,
        "column": 21,
        "endLine": 12,
        "endColumn": 38,
        "path": "poc\\object_tracker\\opencv_mwe_pylint_err.py",
        "symbol": "no-member",
        "message": "Module 'cv2' has no 'RETR_EXTERNAL' member",
        "message-id": "E1101"
    },
    {
        "type": "error",
        "module": "object_tracker.opencv_mwe_pylint_err",
        "obj": "",
        "line": 12,
        "column": 40,
        "endLine": 12,
        "endColumn": 63,
        "path": "poc\\object_tracker\\opencv_mwe_pylint_err.py",
        "symbol": "no-member",
        "message": "Module 'cv2' has no 'CHAIN_APPROX_SIMPLE' member",
        "message-id": "E1101"
    },
    {
        "type": "error",
        "module": "object_tracker.opencv_mwe_pylint_err",
        "obj": "",
        "line": 15,
        "column": 0,
        "endLine": 15,
        "endColumn": 21,
        "path": "poc\\object_tracker\\opencv_mwe_pylint_err.py",
        "symbol": "no-member",
        "message": "Module 'cv2' has no 'destroyAllWindows' member",
        "message-id": "E1101"
    }
]

### Expected behavior

PyLint should recognize all members in the cv2 module.

### Pylint version

pylint 3.2.7
astroid 3.2.4
Python 3.12.7

### OS / Environment

Windows 11
Ubuntu 24.04
Raspberry Pi OS with desktop
Release date: July 4th 2024
System: 32-bit
Kernel version: 6.6
Debian version: 12 (bookworm)

### Additional dependencies

opencv-python==4.10.0.84

**Comment by LambdaScorpii** [comment]
# Edit

Tested it without being an extension, by directly installing the latest version. Seems to cause the same issue, though:
```console
pylint --version
```               
pylint 3.3.1
astroid 3.3.5
Python 3.12.7 (tags/v3.12.7:0b05ead, Oct  1 2024, 03:06:41) [MSC v.1941 64 bit (AMD64)]

```console
pylint opencv_mwe_pylint_err.py
```
************* Module object_tracker.opencv_mwe_pylint_err
opencv_mwe_pylint_err.py:5:6: E1101: Module 'cv2' has no 'VideoCapture' member (no-member)
opencv_mwe_pylint_err.py:6:25: E1101: Module 'cv2' has no 'createBackgroundSubtractorMOG2' member (no-member)
opencv_mwe_pylint_err.py:10:22: E1101: Module 'cv2' has no 'findContours' member (no-member)
opencv_mwe_pylint_err.py:11:21: E1101: Module 'cv2' has no 'RETR_EXTERNAL' member (no-member)
opencv_mwe_pylint_err.py:11:40: E1101: Module 'cv2' has no 'CHAIN_APPROX_SIMPLE' member (no-member)
opencv_mwe_pylint_err.py:14:0: E1101: Module 'cv2' has no 'destroyAllWindows' member (no-member)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)


```python
"""Minimum Working Example showing the issue that PyLint does not recognize cv2 Members"""

import cv2

cap = cv2.VideoCapture("test.mp4")
background_substractor = cv2.createBackgroundSubtractorMOG2()

ret, frame = cap.read()
foreground_mask = background_substractor.apply(frame)
contours, hierarchy = cv2.findContours(
    foreground_mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE
)

cv2.destroyAllWindows()
cap.release()

```

**Comment by ruro** [comment]
Still relevant with `pylint==3.3.6` and `opencv-python==4.11.0.86`.

I think that this is happening, because `opencv` does dynamic loading in `__init__.py` (mutating `globals()`).

Modern versions of `opencv` ship with `pyi` stubs exactly for this purpose. AFAIK, modern versions of `pylint` should be capable of reading type stubs. Unfortunately, this functionality is currently kind of broken.

You can technically "fix" `opencv` linting by running `pylint` with `--prefer-stubs=yes`, however this will break linting for other packages (notably, `numpy`), which makes this solution basically useless. 😦 

Here are some relevant issues:
#9632
#9889

pylint-dev/astroid#2375
#4813
#10017

**Comment by Pierre-Sassoulas** [comment]
Thank you for the investigation @ruro. There's a high probability that this is going to be fixed with proper pyi/stubs/typing handling in pylint and will never be fixed until that happen, so I'm going to close in favor of #4813 / #9889

**Comment by LambdaScorpii** [comment]
Hi @Pierre-Sassoulas and @ruro 

Thanks a lot for your answers, that helped me to understand the issue.

I am quite new to the Open Source World, is there anything I could help with fixing or implementing this?

Cheers

**Comment by Pierre-Sassoulas** [comment]
No problem @LambdaScorpii. I think this is the kind of issue that will require a massive time investment, most likely by a frequent contributor or maintainer. Making the pyi works properly when the code is also available would be the easier first step (if easier because if there's a stub then the inference should use the stub instead of inferring).

**Comment by hraftery** [comment]
Note this is a duplicate of the long running https://github.com/pylint-dev/pylint/issues/2426 and the equivalent on opencv here: https://github.com/opencv/opencv/issues/26344 .

The workaround has evolved and hard to track. For reference, after trying various alternatives, I found adding this to my `.pylintrc` worked for me today:

```
generated-members=cv2
```

Note I didn't find the `.*` necessary. According to the docs that only suppresses E1101 itself, so I'm not too bothered by the concerns raised in the linked issue.
